# Phase 2 — Multi-Seed, Multi-Encoder Training Matrix

Chapter 4 currently reports a **single run** of a single encoder. A single run gives no
handle on seed variance, so it cannot support a claim that one encoder beats another. This
notebook replaces the point estimates with mean +/- sd over three seeds per configuration
and settles the encoder question on our own data.

**The encoder question.** A panelist's published work reports XLNet outperforming BERT on
privacy-policy classification; the thesis claims legal-domain pretraining dominates. Both
claims are about *other* corpora. Running all four encoders through an identical protocol
on identical splits is the only way to know which holds here.

## Run matrix

| Axis | Values |
| --- | --- |
| Encoders (dual-head) | `nlpaueb/legal-bert-base-uncased`, `bert-base-uncased`, `xlnet-base-cased`, `roberta-base` |
| Seeds | 42, 1337, 2024 |
| Head ablation (**all four encoders**) | topic-only, risk-only (dual-head reuses the runs above) |

**12 dual-head runs + 24 ablation runs = 36 (4 encoders x 3 head configs x 3 seeds).**

The ablation was originally run for legal-bert only, on the argument that "does joint
training help?" is a question about the architecture, not the backbone. That was an
assumption, not a measurement, and compute is no longer the binding constraint, so it is
now tested directly: the ablation is run for all four encoders and the dual-vs-single-head
delta is compared across them. XLNet is the encoder most likely to break the pattern — its
seed-to-seed sd is 5-10x every BERT-family encoder's and it pools its summary token from
the last position rather than the first — so an "architecture, not backbone" claim has to
survive XLNet to be worth making. The cross-encoder consistency check at the end of the
notebook is the deliverable; the extra 18 runs only exist to feed it.

## What is held fixed

Everything except encoder / seed / head mode, and it is held fixed by construction — the
runner imports `scripts/lawgic_train_matrix.py`, which reads the same persisted seed-42
split file, the same taxonomy, the same masked-BCE + masked-CE losses (copied line for
line from the original `DualHeadTrainer`), lr 3e-5, batch 8, up to 20 epochs, early
stopping patience 3, weight decay 0.01, warmup 0.06, FP16 on CUDA, max_length 256, and the
same pre-training degenerate-model assertion (a zero-logit model must score topic macro-F1
below 0.95).

## How the pooled representation is chosen per architecture

The two linear heads read one vector per clause. Which token that vector comes from is
**not** the same across these four encoders, and getting it wrong silently cripples a
model rather than erroring:

- **BERT, Legal-BERT, RoBERTa** — the sequence summary is the **first** token
  (`[CLS]` / `<s>`), placed there during pretraining.
- **XLNet** — XLNet is trained with the summary token **appended at the end**. Reading
  position 0 would hand the head an ordinary content token. So XLNet uses the **last**
  token.

`pooled_representation()` in `scripts/lawgic_train_matrix.py` is the single place this
lives. It selects by attention mask rather than by fixed index (`attention_mask.argmax(1)`
for first, `L - 1 - flip(mask).argmax(1)` for last), because XLNet's tokenizer pads on the
**left** while the BERT-family tokenizers pad on the right — a hardcoded `[:, 0]` or
`[:, -1]` would read padding for one of them.

Two further per-architecture quirks are handled in the same adapter, not scattered around:

- **RoBERTa has no `token_type_ids`.** The collator keeps only the keys in
  `tokenizer.model_input_names`, so each tokenizer declares its own contract and no
  `if roberta:` branch is needed anywhere.
- **XLNet's tokenizer needs `sentencepiece`.** Already present in
  `notebooks/requirements.txt` (`sentencepiece==0.2.1`); listed as a manual check below.

**Deviation to record in the manuscript.** The original v3 checkpoint fed the heads BERT's
`pooler_output` (a dense+tanh layer on top of `[CLS]`). The matrix uses the raw first
token instead, for all encoders. Reason: `roberta-base` ships with a *randomly initialised*
pooler, so keeping `pooler_output` would have handicapped RoBERTa for reasons unrelated to
the encoder itself. Consistency across the four arms matters more than bit-matching the
old run, so the legal-bert/seed-42 cell of this matrix is **not** expected to reproduce the
v3 checkpoint exactly — treat the matrix as internally comparable and the Phase 1 numbers
as the checkpoint's own.

In [1]:
import sys
from pathlib import Path


def find_project_root(start: Path) -> Path:
    sentinel = Path("generated_files/lawgic_taxonomy/lawgic_multihead_wide.csv")
    for candidate in (start, *start.parents):
        if (candidate / sentinel).exists():
            return candidate
    raise FileNotFoundError("Run this notebook from inside the lawgic repository.")


PROJECT_ROOT = find_project_root(Path.cwd().resolve())
sys.path.insert(0, str(PROJECT_ROOT / "scripts"))

import json

import numpy as np
import pandas as pd

import lawgic_eval_core as core
import lawgic_train_matrix as tm

pd.set_option("display.width", 160)

split_path = core.persist_split()
corpus = core.load_corpus()
frames = core.split_frames(corpus)
assert {k: len(v) for k, v in frames.items()} == core.EXPECTED_SPLIT_ROWS
print(f"Split artifact: {split_path}")
print("Rows:", {k: len(v) for k, v in frames.items()})

MATRIX = tm.build_matrix()
print(f"\nConfigured runs: {len(MATRIX)}")
display(pd.DataFrame([{
    "run_id": c.run_id, "encoder": c.encoder_name, "seed": c.seed,
    "heads": c.heads, "selection_metric": c.best_metric_key,
} for c in MATRIX]))

Split artifact: C:\Users\Enrique\Coding Projects\Thesis\lawgic\generated_files\lawgic_taxonomy\splits\split_seed42.csv
Rows: {'train': 21183, 'validation': 2648, 'test': 2648}

Configured runs: 18


,run_id,encoder,seed,heads,selection_metric
0,legal-bert-base-uncased__seed42__dual,nlpaueb/legal-bert-base-uncased,42,dual,topic_macro_f1
1,legal-bert-base-uncased__seed1337__dual,nlpaueb/legal-bert-base-uncased,1337,dual,topic_macro_f1
2,legal-bert-base-uncased__seed2024__dual,nlpaueb/legal-bert-base-uncased,2024,dual,topic_macro_f1
3,bert-base-uncased__seed42__dual,bert-base-uncased,42,dual,topic_macro_f1
4,bert-base-uncased__seed1337__dual,bert-base-uncased,1337,dual,topic_macro_f1
5,bert-base-uncased__seed2024__dual,bert-base-uncased,2024,dual,topic_macro_f1
6,xlnet-base-cased__seed42__dual,xlnet-base-cased,42,dual,topic_macro_f1
7,xlnet-base-cased__seed1337__dual,xlnet-base-cased,1337,dual,topic_macro_f1
8,xlnet-base-cased__seed2024__dual,xlnet-base-cased,2024,dual,topic_macro_f1
9,roberta-base__seed42__dual,roberta-base,42,dual,topic_macro_f1


### Extending the matrix

`build_matrix()` returns the original 18 runs (4 encoders x 3 seeds dual-head, plus
legal-bert topic-only/risk-only x 3 seeds). `MATRIX` is a list of `RunConfig` dataclasses,
so extra arms are appended here rather than by rewriting `build_matrix()` — the function
stays the record of what Phase 2 originally ran.

The cell below appends the **18 missing ablation runs**: topic-only and risk-only, three
seeds each, for BERT, XLNet and RoBERTa. `RunConfig.best_metric_key` gives risk-only runs
`risk_macro_f1` and everything else `topic_macro_f1` automatically, so the selection
asymmetry that the legal-bert ablation already uses is replicated for the new encoders by
construction, not by hand. Nothing else about the protocol is touched.

The commented line keeps the earlier five-seed option available. Do **not** add seeds to
only some arms and then compare sds across arms — the sd of 5 draws is not comparable to
the sd of 3.

In [ ]:
MATRIX += [
    tm.RunConfig(encoder_name=encoder, seed=seed, heads=heads)
    for encoder in tm.ENCODERS[1:]
    for heads in ("topic", "risk")
    for seed in tm.SEEDS
]
assert len(MATRIX) == 36 and len({c.run_id for c in MATRIX}) == 36, len(MATRIX)
print(f"Runs after extension: {len(MATRIX)}")
display(pd.DataFrame([{
    "run_id": c.run_id, "encoder": c.encoder_name, "seed": c.seed,
    "heads": c.heads, "selection_metric": c.best_metric_key,
} for c in MATRIX if c.heads != "dual"]))

# MATRIX += [tm.RunConfig(encoder_name=tm.ENCODERS[0], seed=s, heads="dual") for s in (7, 2718)]


## MANUAL STEP — before running the matrix

1. **Model downloads.** The first run of each encoder pulls weights from the HuggingFace
   hub (~440 MB each for `bert-base-uncased`, `xlnet-base-cased`, `roberta-base`;
   legal-bert is already local). Requires network access on the training machine. The
   cell below pre-fetches all three in-notebook via `AutoModel`/`AutoTokenizer`.
2. **`sentencepiece`** must be importable for the XLNet tokenizer. It is already in
   `notebooks/requirements.txt`; the check cell below verifies it rather than installing it.
3. **GPU.** These are 36 full fine-tunes. On CPU this is days, not hours — run on the CUDA
   machine that produced the v3 checkpoint. FP16 switches on automatically on CUDA and off
   elsewhere, matching the original protocol.
4. **Disk.** Each run keeps one checkpoint (`save_total_limit=1`), ~440 MB, plus a small
   `test_logits.npz`. Budget ~20 GB for the full matrix under
   `generated_files/lawgic_taxonomy/runs/`.

Nothing here writes to `saved_models/`; the deployed v3 checkpoint is never touched.

In [5]:
from transformers import AutoModel, AutoTokenizer

MODELS_TO_PREFETCH = ["bert-base-uncased", "xlnet-base-cased", "roberta-base"]

for model_name in MODELS_TO_PREFETCH:
    print(f"Downloading {model_name} ...")
    AutoTokenizer.from_pretrained(model_name)
    AutoModel.from_pretrained(model_name)
    print(f"  done: {model_name}")

print("\nAll three encoders cached locally.")

  done: bert-base-uncased
  done: xlnet-base-cased


Some weights of RobertaModel were not initialized from the model checkpoint at roberta-base and are newly initialized: ['pooler.dense.bias', 'pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


  done: roberta-base

All three encoders cached locally.


In [9]:
%conda install conda-forge::sentencepiece

3 channel Terms of Service accepted
Channels:
 - defaults
 - conda-forge
Platform: win-64
Solving environment: done

## Package Plan ##

  environment location: c:\Users\Enrique\anaconda3\envs\thesis-env

  added / updated specs:
    - conda-forge::sentencepiece


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    libabseil-20260526.0       | cxx17_h4cdcee1_0         1.9 MB
    libprotobuf-7.35.1         |       hb5abd84_0         6.9 MB
    libsentencepiece-0.2.1     |       h1e80020_4         1.4 MB  conda-forge
    sentencepiece-0.2.1        |       hb9477dd_4          20 KB  conda-forge
    sentencepiece-python-0.2.1 |  py314h2f88111_4         3.2 MB  conda-forge
    sentencepiece-spm-0.2.1    |       h1e80020_4         173 KB  conda-forge
    ------------------------------------------------------------
                                           Total:        13.5 MB

The following NEW 



==> WARNING: A newer version of conda exists. <==
    current version: 25.5.1
    latest version: 26.5.3

Please update conda by running

    $ conda update -n base -c defaults conda




In [10]:
import importlib.util

print("sentencepiece:", "OK" if importlib.util.find_spec("sentencepiece") else "MISSING — XLNet will fail")
print("scipy:", "OK" if importlib.util.find_spec("scipy") else "MISSING — McNemar will fail")

device_label, device = tm.detect_device()
print(f"device: {device_label} (fp16={device_label == 'cuda'})")
if device_label != "cuda":
    print("WARNING: not on CUDA. The matrix will take days. Stop and move to the GPU machine.")

sentencepiece: OK
scipy: OK
device: cuda (fp16=True)


## Expected wall time

The v3 run's `trainer_state.json` records **17 epochs** before early stopping (best at
epoch 14, patience 3) at batch size 8 over 21,183 training rows = 2,648 optimizer steps
per epoch, and an eval throughput of ~455 clauses/s on the original CUDA device. It does
**not** record `train_runtime` — the notebook that produced it never logged the summary —
so the per-run wall time must be **measured on the first run**, not assumed.

The cell below prints the derived lower bound from what *is* recorded, then the runner
stores the real `wall_seconds` for every run. After the first run completes, multiply.

In [ ]:
state_path = core.CHECKPOINT_DIR / "checkpoints/checkpoint-45016/trainer_state.json"
if state_path.exists():
    state = json.loads(state_path.read_text())
    evals = [h for h in state["log_history"] if "eval_runtime" in h]
    eval_throughput = float(np.mean([h["eval_samples_per_second"] for h in evals]))
    epochs = float(state["epoch"])
    # Training is roughly 3-4x the cost of inference per sample (forward + backward + optimizer).
    optimistic_seconds = epochs * (len(frames["train"]) / (eval_throughput / 3.5))
    print(f"v3 run: {epochs:.0f} epochs, eval throughput {eval_throughput:.0f} clauses/s")
    print(f"Derived LOWER BOUND per run: ~{optimistic_seconds / 60:.0f} min "
          f"-> ~{len(MATRIX) * optimistic_seconds / 3600:.1f} h for {len(MATRIX)} runs")
    print("This is an extrapolation, not a measurement. Trust wall_seconds from run 1 instead.")
else:
    print("No v3 trainer_state.json found; wall time must be measured on the first run.")

## Runner

Each config trains, evaluates on the frozen test split, and writes to
`generated_files/lawgic_taxonomy/runs/<run_id>/`:

- `metrics.json` — config + headline test metrics + `wall_seconds` + `epochs_run`
- `test_logits.npz` — test logits, labels and masks (so aggregation, bootstrap and paired
  tests never need to re-run inference)
- `per_topic.csv` — per-topic precision / recall / F1 / support

Completed runs are skipped, so the cell is **resumable**: interrupt it, restart the kernel,
re-run. Set `FORCE_RERUN = True` to redo everything.

In [12]:
FORCE_RERUN = False

records = []
for index, config in enumerate(MATRIX, start=1):
    target = tm.RUNS_DIR / config.run_id / "metrics.json"
    if target.exists() and not FORCE_RERUN:
        print(f"[{index}/{len(MATRIX)}] skip {config.run_id} (already complete)")
        records.append(json.loads(target.read_text()))
        continue
    print(f"[{index}/{len(MATRIX)}] running {config.run_id} ...")
    records.append(tm.run_config(config))

print(f"\nCompleted {len(records)} runs.")

[1/18] running legal-bert-base-uncased__seed42__dual ...


You're using a BertTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.


Epoch,Training Loss,Validation Loss,Topic Macro F1,Topic Micro F1,Topic Weighted F1,Topic Observed Positions,Risk Accuracy,Risk Macro F1,Risk Weighted F1,Risk Valid Rows
1,0.769300,0.775791,0.064763,0.211135,0.171681,96620.000000,0.804003,0.797135,0.803743,2648.000000
2,0.604600,0.690023,0.381158,0.631470,0.574289,96620.000000,0.814955,0.805029,0.814502,2648.000000
3,0.375200,0.640507,0.603367,0.755805,0.733281,96620.000000,0.837991,0.832059,0.836521,2648.000000
4,0.388900,0.719763,0.642261,0.782893,0.770978,96620.000000,0.841390,0.835409,0.839652,2648.000000
5,0.249500,0.825951,0.669976,0.803822,0.795191,96620.000000,0.843656,0.839919,0.844028,2648.000000
6,0.146400,0.876684,0.704711,0.824830,0.818666,96620.000000,0.848565,0.843226,0.848650,2648.000000
7,0.138400,1.025527,0.710526,0.829706,0.825624,96620.000000,0.853097,0.846846,0.852055,2648.000000
8,0.092200,1.074190,0.732378,0.832284,0.829008,96620.000000,0.845544,0.840759,0.845042,2648.000000
9,0.046300,1.102909,0.744242,0.838182,0.835651,96620.000000,0.848565,0.843589,0.848294,2648.000000
10,0.095000,1.162956,0.739702,0.837367,0.835221,96620.000000,0.847810,0.842722,0.847207,2648.000000


[legal-bert-base-uncased__seed42__dual] 63.8 min | topic_macro_f1=0.7741 topic_micro_f1=0.8363 risk_accuracy=0.8406 risk_macro_f1=0.8349
[2/18] running legal-bert-base-uncased__seed1337__dual ...


You're using a BertTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.


Epoch,Training Loss,Validation Loss,Topic Macro F1,Topic Micro F1,Topic Weighted F1,Topic Observed Positions,Risk Accuracy,Risk Macro F1,Risk Weighted F1,Risk Valid Rows
1,0.665100,0.841575,0.098436,0.292189,0.237333,96620.000000,0.786631,0.776126,0.786774,2648.000000
2,0.571300,0.659903,0.437594,0.649453,0.592348,96620.000000,0.816843,0.813411,0.817000,2648.000000
3,0.442800,0.704480,0.572639,0.737571,0.709921,96620.000000,0.835725,0.828719,0.835259,2648.000000
4,0.263400,0.731843,0.650250,0.790295,0.777764,96620.000000,0.841767,0.835669,0.841540,2648.000000
5,0.275800,0.765842,0.668034,0.804688,0.792548,96620.000000,0.839502,0.832701,0.838285,2648.000000
6,0.287000,0.880332,0.689551,0.820658,0.815068,96620.000000,0.847432,0.842175,0.846934,2648.000000
7,0.183700,1.001644,0.729665,0.829157,0.825064,96620.000000,0.845921,0.844195,0.846698,2648.000000
8,0.069500,1.059433,0.732081,0.836083,0.831635,96620.000000,0.853474,0.850188,0.853387,2648.000000
9,0.061700,1.115113,0.770266,0.835931,0.833880,96620.000000,0.858006,0.854155,0.858445,2648.000000
10,0.089100,1.186666,0.756061,0.836447,0.834530,96620.000000,0.852341,0.847797,0.852053,2648.000000


[legal-bert-base-uncased__seed1337__dual] 45.3 min | topic_macro_f1=0.7693 topic_micro_f1=0.8329 risk_accuracy=0.8380 risk_macro_f1=0.8329
[3/18] running legal-bert-base-uncased__seed2024__dual ...


You're using a BertTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.


Epoch,Training Loss,Validation Loss,Topic Macro F1,Topic Micro F1,Topic Weighted F1,Topic Observed Positions,Risk Accuracy,Risk Macro F1,Risk Weighted F1,Risk Valid Rows
1,0.707900,0.842565,0.091213,0.295473,0.229743,96620.000000,0.786631,0.772659,0.784099,2648.000000
2,0.555800,0.630114,0.416257,0.646568,0.586441,96620.000000,0.823640,0.818654,0.824256,2648.000000
3,0.369000,0.668306,0.596066,0.758979,0.735614,96620.000000,0.840257,0.833411,0.839364,2648.000000
4,0.433300,0.674484,0.642097,0.788335,0.777910,96620.000000,0.854230,0.849928,0.853976,2648.000000
5,0.236800,0.826601,0.672423,0.815573,0.804810,96620.000000,0.843278,0.839655,0.843049,2648.000000
6,0.174900,0.861291,0.690868,0.823320,0.815670,96620.000000,0.849320,0.844385,0.849194,2648.000000
7,0.214800,1.017144,0.710336,0.825485,0.820678,96620.000000,0.844411,0.838652,0.843013,2648.000000
8,0.107400,1.041995,0.724654,0.836184,0.833136,96620.000000,0.853097,0.847730,0.852033,2648.000000
9,0.064300,1.138210,0.745106,0.839823,0.837688,96620.000000,0.851586,0.846345,0.851036,2648.000000
10,0.026100,1.177273,0.746205,0.834559,0.832681,96620.000000,0.848187,0.842246,0.847226,2648.000000


[legal-bert-base-uncased__seed2024__dual] 147.9 min | topic_macro_f1=0.7703 topic_micro_f1=0.8314 risk_accuracy=0.8365 risk_macro_f1=0.8309
[4/18] running bert-base-uncased__seed42__dual ...


You're using a BertTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.


Epoch,Training Loss,Validation Loss,Topic Macro F1,Topic Micro F1,Topic Weighted F1,Topic Observed Positions,Risk Accuracy,Risk Macro F1,Risk Weighted F1,Risk Valid Rows
1,0.806800,0.823286,0.063201,0.194485,0.163105,96620.000000,0.795695,0.786851,0.796567,2648.000000
2,0.621400,0.663765,0.380761,0.630852,0.571110,96620.000000,0.824773,0.818124,0.824873,2648.000000
3,0.370500,0.672070,0.582540,0.743960,0.725051,96620.000000,0.833837,0.828313,0.833340,2648.000000
4,0.347100,0.795403,0.630396,0.770891,0.755749,96620.000000,0.830816,0.823401,0.829067,2648.000000
5,0.268000,0.874795,0.675349,0.804150,0.795324,96620.000000,0.843656,0.838445,0.843282,2648.000000
6,0.185900,1.017555,0.697971,0.828317,0.821082,96620.000000,0.843278,0.837787,0.842802,2648.000000
7,0.081400,1.112253,0.715856,0.829842,0.825247,96620.000000,0.851964,0.848259,0.851858,2648.000000
8,0.081100,1.232278,0.733556,0.831126,0.827956,96620.000000,0.842900,0.837849,0.841586,2648.000000
9,0.042100,1.306083,0.734161,0.834444,0.830743,96620.000000,0.840257,0.835832,0.840437,2648.000000
10,0.038400,1.365361,0.740190,0.826128,0.823185,96620.000000,0.841767,0.837065,0.842064,2648.000000


[bert-base-uncased__seed42__dual] 63.7 min | topic_macro_f1=0.7757 topic_micro_f1=0.8329 risk_accuracy=0.8263 risk_macro_f1=0.8207
[5/18] running bert-base-uncased__seed1337__dual ...


You're using a BertTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.


Epoch,Training Loss,Validation Loss,Topic Macro F1,Topic Micro F1,Topic Weighted F1,Topic Observed Positions,Risk Accuracy,Risk Macro F1,Risk Weighted F1,Risk Valid Rows
1,0.695800,0.851256,0.069894,0.198656,0.164459,96620.000000,0.779834,0.766365,0.778264,2648.000000
2,0.532000,0.647787,0.380868,0.630589,0.577474,96620.000000,0.814955,0.809445,0.814426,2648.000000
3,0.373000,0.675959,0.574601,0.741274,0.716915,96620.000000,0.834592,0.828786,0.834501,2648.000000
4,0.274600,0.736560,0.638751,0.788491,0.775683,96620.000000,0.845166,0.840899,0.844893,2648.000000
5,0.208600,0.929713,0.666763,0.807635,0.795476,96620.000000,0.839502,0.834108,0.839221,2648.000000
6,0.167800,0.974446,0.687263,0.818556,0.812730,96620.000000,0.838746,0.833223,0.837892,2648.000000
7,0.182200,1.044780,0.718129,0.825118,0.820772,96620.000000,0.844411,0.840857,0.844408,2648.000000
8,0.059200,1.188978,0.720049,0.827652,0.823314,96620.000000,0.852341,0.848104,0.852287,2648.000000
9,0.076300,1.234349,0.734293,0.827990,0.825103,96620.000000,0.852719,0.847547,0.852335,2648.000000
10,0.046000,1.309831,0.740156,0.821966,0.818943,96620.000000,0.846299,0.840751,0.845857,2648.000000


[bert-base-uncased__seed1337__dual] 49.2 min | topic_macro_f1=0.7681 topic_micro_f1=0.8336 risk_accuracy=0.8263 risk_macro_f1=0.8196
[6/18] running bert-base-uncased__seed2024__dual ...


You're using a BertTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.


Epoch,Training Loss,Validation Loss,Topic Macro F1,Topic Micro F1,Topic Weighted F1,Topic Observed Positions,Risk Accuracy,Risk Macro F1,Risk Weighted F1,Risk Valid Rows
1,0.654400,0.858890,0.054167,0.168853,0.141899,96620.000000,0.790785,0.779747,0.788839,2648.000000
2,0.563400,0.636156,0.373500,0.614317,0.561237,96620.000000,0.835725,0.830549,0.835129,2648.000000
3,0.370200,0.755034,0.583331,0.748039,0.726075,96620.000000,0.814199,0.805083,0.812752,2648.000000
4,0.410500,0.804622,0.641292,0.786711,0.775899,96620.000000,0.832704,0.827817,0.832387,2648.000000
5,0.271500,0.865902,0.668346,0.807051,0.796859,96620.000000,0.838746,0.832315,0.838691,2648.000000
6,0.121300,0.970835,0.689974,0.819843,0.812686,96620.000000,0.841767,0.838357,0.842396,2648.000000
7,0.148000,1.157131,0.712942,0.827180,0.822476,96620.000000,0.833082,0.828526,0.833288,2648.000000
8,0.068300,1.267551,0.738705,0.827692,0.825633,96620.000000,0.841012,0.836502,0.840095,2648.000000
9,0.038600,1.320127,0.735786,0.822094,0.818885,96620.000000,0.831949,0.827337,0.831591,2648.000000
10,0.032200,1.400717,0.728509,0.828972,0.826234,96620.000000,0.838369,0.833649,0.838023,2648.000000


[bert-base-uncased__seed2024__dual] 64.6 min | topic_macro_f1=0.7803 topic_micro_f1=0.8368 risk_accuracy=0.8365 risk_macro_f1=0.8312
[7/18] running xlnet-base-cased__seed42__dual ...


You're using a XLNetTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.


Epoch,Training Loss,Validation Loss,Topic Macro F1,Topic Micro F1,Topic Weighted F1,Topic Observed Positions,Risk Accuracy,Risk Macro F1,Risk Weighted F1,Risk Valid Rows
1,0.754700,0.729617,0.202748,0.441302,0.387346,96620.000000,0.796450,0.788246,0.796460,2648.000000
2,0.656400,0.716310,0.535690,0.699789,0.678695,96620.000000,0.818731,0.810835,0.818564,2648.000000
3,0.406200,0.718326,0.632639,0.769748,0.757547,96620.000000,0.835725,0.829014,0.835046,2648.000000
4,0.414600,0.731784,0.669052,0.784550,0.774253,96620.000000,0.837991,0.834018,0.837403,2648.000000
5,0.371600,0.854279,0.703109,0.806431,0.800931,96620.000000,0.843278,0.840263,0.843175,2648.000000
6,0.245000,1.029569,0.719500,0.814000,0.809738,96620.000000,0.840257,0.837671,0.840649,2648.000000
7,0.222100,1.114610,0.725532,0.822910,0.818564,96620.000000,0.848943,0.844108,0.848164,2648.000000
8,0.140000,1.314959,0.714639,0.821663,0.817322,96620.000000,0.845166,0.841921,0.843909,2648.000000
9,0.141600,1.412801,0.724853,0.828539,0.826568,96620.000000,0.847810,0.842790,0.847404,2648.000000
10,0.126100,1.497892,0.745826,0.829689,0.826520,96620.000000,0.852341,0.849291,0.852095,2648.000000


[xlnet-base-cased__seed42__dual] 73.5 min | topic_macro_f1=0.7681 topic_micro_f1=0.8311 risk_accuracy=0.8338 risk_macro_f1=0.8274
[8/18] running xlnet-base-cased__seed1337__dual ...


You're using a XLNetTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.


Epoch,Training Loss,Validation Loss,Topic Macro F1,Topic Micro F1,Topic Weighted F1,Topic Observed Positions,Risk Accuracy,Risk Macro F1,Risk Weighted F1,Risk Valid Rows
1,0.703900,0.787621,0.317944,0.543387,0.493599,96620.000000,0.790785,0.777469,0.789650,2648.000000
2,0.550900,0.693884,0.554190,0.712929,0.684919,96620.000000,0.813066,0.806273,0.813196,2648.000000
3,0.419000,0.631152,0.631415,0.774445,0.762601,96620.000000,0.831571,0.829030,0.831738,2648.000000
4,0.386400,0.737441,0.671922,0.792062,0.784913,96620.000000,0.839124,0.833695,0.838603,2648.000000
5,0.293900,0.906648,0.690062,0.801050,0.793892,96620.000000,0.834592,0.830004,0.834897,2648.000000
6,0.321800,0.958380,0.711874,0.820396,0.814926,96620.000000,0.835347,0.829594,0.834863,2648.000000
7,0.293200,1.167251,0.734564,0.826283,0.824284,96620.000000,0.835347,0.831279,0.835502,2648.000000
8,0.201800,1.292096,0.738682,0.823709,0.820764,96620.000000,0.841012,0.836500,0.840410,2648.000000
9,0.107600,1.293092,0.759246,0.828903,0.827507,96620.000000,0.836858,0.831010,0.835819,2648.000000
10,0.081100,1.507189,0.739730,0.829238,0.827096,96620.000000,0.839879,0.836663,0.839907,2648.000000


[xlnet-base-cased__seed1337__dual] 64.2 min | topic_macro_f1=0.7283 topic_micro_f1=0.8290 risk_accuracy=0.8297 risk_macro_f1=0.8231
[9/18] running xlnet-base-cased__seed2024__dual ...


You're using a XLNetTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.


Epoch,Training Loss,Validation Loss,Topic Macro F1,Topic Micro F1,Topic Weighted F1,Topic Observed Positions,Risk Accuracy,Risk Macro F1,Risk Weighted F1,Risk Valid Rows
1,0.701300,0.897286,0.275448,0.549600,0.476805,96620.000000,0.778701,0.763330,0.776090,2648.000000
2,0.552700,0.640958,0.507782,0.681004,0.647342,96620.000000,0.817221,0.813058,0.818318,2648.000000
3,0.428300,0.672969,0.643282,0.771068,0.762393,96620.000000,0.831949,0.824744,0.830734,2648.000000
4,0.537300,0.715558,0.661792,0.792469,0.781829,96620.000000,0.836480,0.828311,0.835564,2648.000000
5,0.386200,0.806338,0.708235,0.805951,0.802289,96620.000000,0.829683,0.828054,0.831009,2648.000000
6,0.279400,0.912202,0.727946,0.820513,0.816024,96620.000000,0.836858,0.834710,0.837434,2648.000000
7,0.255000,1.135440,0.745701,0.826879,0.824571,96620.000000,0.836103,0.831658,0.836111,2648.000000
8,0.143200,1.224459,0.722946,0.819788,0.817982,96620.000000,0.842523,0.838110,0.842220,2648.000000
9,0.141800,1.351844,0.740278,0.828521,0.826559,96620.000000,0.846677,0.843211,0.846685,2648.000000
10,0.195500,1.389906,0.760251,0.828223,0.826334,96620.000000,0.840257,0.835700,0.839878,2648.000000


[xlnet-base-cased__seed2024__dual] 107.1 min | topic_macro_f1=0.7892 topic_micro_f1=0.8419 risk_accuracy=0.8433 risk_macro_f1=0.8382
[10/18] running roberta-base__seed42__dual ...


Some weights of RobertaModel were not initialized from the model checkpoint at roberta-base and are newly initialized: ['pooler.dense.bias', 'pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
You're using a RobertaTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.


Epoch,Training Loss,Validation Loss,Topic Macro F1,Topic Micro F1,Topic Weighted F1,Topic Observed Positions,Risk Accuracy,Risk Macro F1,Risk Weighted F1,Risk Valid Rows
1,0.794800,0.755354,0.132533,0.379187,0.302952,96620.000000,0.807402,0.797433,0.806755,2648.000000
2,0.624500,0.787349,0.394628,0.622651,0.562968,96620.000000,0.795317,0.781417,0.793332,2648.000000
3,0.434000,0.705147,0.594460,0.745872,0.727694,96620.000000,0.826662,0.818616,0.824917,2648.000000
4,0.412200,0.710568,0.624079,0.760824,0.746332,96620.000000,0.837613,0.830811,0.835825,2648.000000
5,0.392200,0.765190,0.656811,0.790253,0.781550,96620.000000,0.842523,0.837148,0.842172,2648.000000
6,0.289400,0.889759,0.673184,0.803189,0.796724,96620.000000,0.848187,0.843526,0.847763,2648.000000
7,0.249300,0.926642,0.697654,0.814610,0.809962,96620.000000,0.842900,0.836729,0.841501,2648.000000
8,0.128800,1.029624,0.693498,0.817344,0.812271,96620.000000,0.838369,0.832866,0.837204,2648.000000
9,0.124700,1.071074,0.749232,0.829606,0.827154,96620.000000,0.842523,0.836014,0.841555,2648.000000
10,0.123100,1.198429,0.734720,0.829853,0.826871,96620.000000,0.835725,0.830836,0.835442,2648.000000


[roberta-base__seed42__dual] 68.6 min | topic_macro_f1=0.7781 topic_micro_f1=0.8339 risk_accuracy=0.8372 risk_macro_f1=0.8323
[11/18] running roberta-base__seed1337__dual ...


Some weights of RobertaModel were not initialized from the model checkpoint at roberta-base and are newly initialized: ['pooler.dense.bias', 'pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
You're using a RobertaTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.


Epoch,Training Loss,Validation Loss,Topic Macro F1,Topic Micro F1,Topic Weighted F1,Topic Observed Positions,Risk Accuracy,Risk Macro F1,Risk Weighted F1,Risk Valid Rows
1,0.652700,0.879982,0.055645,0.168747,0.140524,96620.000000,0.782100,0.768276,0.780247,2648.000000
2,0.562400,0.719881,0.456537,0.649528,0.604102,96620.000000,0.817598,0.813459,0.818244,2648.000000
3,0.476500,0.627245,0.584604,0.743457,0.727297,96620.000000,0.836103,0.830983,0.835717,2648.000000
4,0.444800,0.693747,0.640807,0.780354,0.768372,96620.000000,0.839124,0.834677,0.839066,2648.000000
5,0.319500,0.733140,0.658247,0.797388,0.785505,96620.000000,0.850076,0.843424,0.849559,2648.000000
6,0.349100,0.756211,0.677134,0.801204,0.795532,96620.000000,0.845166,0.840441,0.845093,2648.000000
7,0.212600,0.941550,0.704065,0.817814,0.812604,96620.000000,0.829683,0.825762,0.830146,2648.000000
8,0.157700,0.975177,0.720246,0.826496,0.821643,96620.000000,0.848187,0.843699,0.847859,2648.000000
9,0.167100,1.103754,0.721617,0.824352,0.820903,96620.000000,0.842145,0.838000,0.842744,2648.000000
10,0.099300,1.128295,0.742071,0.827522,0.824292,96620.000000,0.846677,0.842368,0.846471,2648.000000


[roberta-base__seed1337__dual] 69.3 min | topic_macro_f1=0.7803 topic_micro_f1=0.8398 risk_accuracy=0.8406 risk_macro_f1=0.8343
[12/18] running roberta-base__seed2024__dual ...


Some weights of RobertaModel were not initialized from the model checkpoint at roberta-base and are newly initialized: ['pooler.dense.bias', 'pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
You're using a RobertaTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.


Epoch,Training Loss,Validation Loss,Topic Macro F1,Topic Micro F1,Topic Weighted F1,Topic Observed Positions,Risk Accuracy,Risk Macro F1,Risk Weighted F1,Risk Valid Rows
1,0.706500,0.948857,0.135064,0.365215,0.300190,96620.000000,0.771903,0.756331,0.768052,2648.000000
2,0.557000,0.653102,0.435255,0.629972,0.588846,96620.000000,0.811178,0.806818,0.812277,2648.000000
3,0.455900,0.686708,0.587671,0.741731,0.720299,96620.000000,0.819864,0.811905,0.817863,2648.000000
4,0.488100,0.721064,0.634502,0.778531,0.768016,96620.000000,0.836103,0.827874,0.834312,2648.000000
5,0.375500,0.734499,0.655797,0.798631,0.788685,96620.000000,0.841012,0.836119,0.841077,2648.000000
6,0.247900,0.787721,0.677625,0.801066,0.794574,96620.000000,0.843278,0.838643,0.843585,2648.000000
7,0.314000,0.902336,0.702080,0.823775,0.818624,96620.000000,0.843656,0.838646,0.843404,2648.000000
8,0.184100,0.983751,0.694514,0.811570,0.808165,96620.000000,0.850453,0.844647,0.849436,2648.000000
9,0.176700,0.993176,0.736696,0.832169,0.830663,96620.000000,0.850831,0.846186,0.850848,2648.000000
10,0.168900,1.084685,0.724861,0.831312,0.829344,96620.000000,0.853097,0.847045,0.851962,2648.000000


[roberta-base__seed2024__dual] 68.4 min | topic_macro_f1=0.7682 topic_micro_f1=0.8375 risk_accuracy=0.8474 risk_macro_f1=0.8422
[13/18] running legal-bert-base-uncased__seed42__topic ...


You're using a BertTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.


Epoch,Training Loss,Validation Loss,Topic Macro F1,Topic Micro F1,Topic Weighted F1,Topic Observed Positions,Risk Accuracy,Risk Macro F1,Risk Weighted F1,Risk Valid Rows
1,0.079000,0.143918,0.529524,0.736195,0.709733,96620.000000,0.466012,0.240371,0.321677,2648.000000
2,0.056600,0.108563,0.666138,0.797020,0.791198,96620.000000,0.459215,0.328923,0.388074,2648.000000
3,0.034800,0.110639,0.714224,0.812986,0.807748,96620.000000,0.456571,0.314321,0.376003,2648.000000
4,0.030700,0.099835,0.734194,0.826263,0.823007,96620.000000,0.468278,0.324005,0.386359,2648.000000
5,0.022200,0.113048,0.756296,0.819066,0.819628,96620.000000,0.458837,0.292707,0.362289,2648.000000
6,0.014000,0.092277,0.769112,0.831516,0.829792,96620.000000,0.461103,0.311242,0.374092,2648.000000
7,0.012800,0.107626,0.777830,0.835104,0.833720,96620.000000,0.474320,0.297455,0.365990,2648.000000
8,0.010100,0.108689,0.763895,0.833025,0.831985,96620.000000,0.441843,0.311710,0.369674,2648.000000
9,0.010000,0.114617,0.761746,0.832384,0.830495,96620.000000,0.446752,0.309914,0.367535,2648.000000
10,0.006200,0.124257,0.759627,0.830409,0.829718,96620.000000,0.461103,0.311522,0.374069,2648.000000


[legal-bert-base-uncased__seed42__topic] 32.2 min | topic_macro_f1=0.7857 topic_micro_f1=0.8329 risk_accuracy=nan risk_macro_f1=nan
[14/18] running legal-bert-base-uncased__seed1337__topic ...


You're using a BertTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.


Epoch,Training Loss,Validation Loss,Topic Macro F1,Topic Micro F1,Topic Weighted F1,Topic Observed Positions,Risk Accuracy,Risk Macro F1,Risk Weighted F1,Risk Valid Rows
1,0.082900,0.145462,0.534904,0.724420,0.689648,96620.000000,0.289653,0.248181,0.215778,2648.000000
2,0.055200,0.115149,0.674380,0.804292,0.793748,96620.000000,0.335347,0.308850,0.292813,2648.000000
3,0.032600,0.102382,0.695338,0.816008,0.810292,96620.000000,0.374622,0.358188,0.349164,2648.000000
4,0.033700,0.099174,0.736771,0.830779,0.826268,96620.000000,0.348565,0.336797,0.324448,2648.000000
5,0.021500,0.100362,0.745344,0.827463,0.825042,96620.000000,0.353474,0.342289,0.331824,2648.000000
6,0.015800,0.103340,0.760818,0.824286,0.823924,96620.000000,0.343278,0.342491,0.328300,2648.000000
7,0.014900,0.100705,0.750901,0.835023,0.833796,96620.000000,0.348943,0.338794,0.330181,2648.000000
8,0.008500,0.115474,0.770100,0.831161,0.829749,96620.000000,0.375755,0.363939,0.359335,2648.000000
9,0.006600,0.129632,0.750055,0.829199,0.829326,96620.000000,0.354607,0.344798,0.333411,2648.000000
10,0.004800,0.120087,0.769177,0.833061,0.832299,96620.000000,0.354230,0.346064,0.339525,2648.000000


[legal-bert-base-uncased__seed1337__topic] 51.2 min | topic_macro_f1=0.7788 topic_micro_f1=0.8377 risk_accuracy=nan risk_macro_f1=nan
[15/18] running legal-bert-base-uncased__seed2024__topic ...


You're using a BertTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.


Epoch,Training Loss,Validation Loss,Topic Macro F1,Topic Micro F1,Topic Weighted F1,Topic Observed Positions,Risk Accuracy,Risk Macro F1,Risk Weighted F1,Risk Valid Rows
1,0.084800,0.141275,0.580587,0.753451,0.731471,96620.000000,0.373489,0.328311,0.271685,2648.000000
2,0.055400,0.102995,0.675080,0.803475,0.793629,96620.000000,0.381042,0.322909,0.264663,2648.000000
3,0.043900,0.101103,0.735255,0.829842,0.825698,96620.000000,0.391239,0.336942,0.276982,2648.000000
4,0.036600,0.099242,0.739206,0.827781,0.823420,96620.000000,0.369713,0.317213,0.262636,2648.000000
5,0.020000,0.101684,0.747051,0.832622,0.830751,96620.000000,0.374245,0.320294,0.264876,2648.000000
6,0.013200,0.102008,0.750866,0.826438,0.825539,96620.000000,0.369335,0.317874,0.261176,2648.000000
7,0.012200,0.097587,0.764043,0.836250,0.833654,96620.000000,0.371979,0.322362,0.269407,2648.000000
8,0.007100,0.105339,0.768907,0.836364,0.835768,96620.000000,0.376511,0.324054,0.267734,2648.000000
9,0.009300,0.124041,0.768960,0.826833,0.827885,96620.000000,0.367069,0.319188,0.264916,2648.000000
10,0.008100,0.124431,0.785321,0.836994,0.837027,96620.000000,0.371601,0.320585,0.265990,2648.000000


[legal-bert-base-uncased__seed2024__topic] 40.9 min | topic_macro_f1=0.7679 topic_micro_f1=0.8295 risk_accuracy=nan risk_macro_f1=nan
[16/18] running legal-bert-base-uncased__seed42__risk ...


You're using a BertTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.


Epoch,Training Loss,Validation Loss,Topic Macro F1,Topic Micro F1,Topic Weighted F1,Topic Observed Positions,Risk Accuracy,Risk Macro F1,Risk Weighted F1,Risk Valid Rows
1,0.609000,0.518894,0.074662,0.081543,0.127277,96620.000000,0.799471,0.789727,0.799658,2648.000000
2,0.479000,0.533805,0.071683,0.080520,0.111815,96620.000000,0.819486,0.808079,0.817766,2648.000000
3,0.321000,0.511072,0.085664,0.094412,0.134372,96620.000000,0.837613,0.829735,0.835892,2648.000000
4,0.338500,0.561137,0.081038,0.090766,0.123832,96620.000000,0.830816,0.823406,0.829705,2648.000000
5,0.212200,0.771932,0.080975,0.090703,0.126051,96620.000000,0.838746,0.832646,0.838792,2648.000000
6,0.166400,0.892199,0.081395,0.090671,0.126999,96620.000000,0.830438,0.827085,0.831239,2648.000000
7,0.147100,1.000898,0.078815,0.089907,0.125229,96620.000000,0.837236,0.831830,0.837565,2648.000000
8,0.043200,1.116724,0.078463,0.089169,0.123178,96620.000000,0.830438,0.824455,0.829624,2648.000000


[legal-bert-base-uncased__seed42__risk] 25.6 min | topic_macro_f1=nan topic_micro_f1=nan risk_accuracy=0.8278 risk_macro_f1=0.8234
[17/18] running legal-bert-base-uncased__seed1337__risk ...


You're using a BertTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.


Epoch,Training Loss,Validation Loss,Topic Macro F1,Topic Micro F1,Topic Weighted F1,Topic Observed Positions,Risk Accuracy,Risk Macro F1,Risk Weighted F1,Risk Valid Rows
1,0.517000,0.570518,0.083146,0.095569,0.134571,96620.000000,0.772281,0.758391,0.772466,2648.000000
2,0.489500,0.486974,0.086604,0.093855,0.137183,96620.000000,0.825151,0.819743,0.824792,2648.000000
3,0.350700,0.533163,0.087033,0.096724,0.147276,96620.000000,0.842523,0.838042,0.842195,2648.000000
4,0.238400,0.683652,0.089286,0.098679,0.153052,96620.000000,0.828927,0.823575,0.829422,2648.000000
5,0.158400,0.787031,0.090478,0.098653,0.149630,96620.000000,0.843656,0.835378,0.842026,2648.000000
6,0.172300,0.954357,0.089879,0.100279,0.152976,96620.000000,0.832704,0.826808,0.832642,2648.000000


[legal-bert-base-uncased__seed1337__risk] 19.2 min | topic_macro_f1=nan topic_micro_f1=nan risk_accuracy=0.8233 risk_macro_f1=0.8173
[18/18] running legal-bert-base-uncased__seed2024__risk ...


You're using a BertTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.


Epoch,Training Loss,Validation Loss,Topic Macro F1,Topic Micro F1,Topic Weighted F1,Topic Observed Positions,Risk Accuracy,Risk Macro F1,Risk Weighted F1,Risk Valid Rows
1,0.526800,0.522216,0.087264,0.093772,0.140698,96620.000000,0.793429,0.786405,0.793565,2648.000000
2,0.459600,0.462695,0.089864,0.103816,0.152229,96620.000000,0.827039,0.821620,0.827467,2648.000000
3,0.310300,0.589685,0.091994,0.099152,0.153241,96620.000000,0.832704,0.825442,0.831539,2648.000000
4,0.397200,0.635954,0.081540,0.090307,0.137275,96620.000000,0.836103,0.829607,0.835689,2648.000000
5,0.228200,0.812985,0.088410,0.092150,0.145271,96620.000000,0.839502,0.835424,0.839773,2648.000000
6,0.176700,0.949717,0.086535,0.089583,0.132480,96620.000000,0.833459,0.828439,0.832977,2648.000000
7,0.278300,1.019065,0.093257,0.094398,0.145490,96620.000000,0.832326,0.828742,0.832454,2648.000000
8,0.071800,1.164697,0.097891,0.099262,0.149093,96620.000000,0.838369,0.833414,0.838078,2648.000000


[legal-bert-base-uncased__seed2024__risk] 25.6 min | topic_macro_f1=nan topic_micro_f1=nan risk_accuracy=0.8191 risk_macro_f1=0.8154

Completed 18 runs.


Save best models to directory

In [19]:
# Save the best (highest validation metric) dual-head model per encoder to
# saved_models/, in the same file layout as lawgic_classifier_legal-bert_v3
# (model_state_dict.pt + encoder/tokenizer + head weights + taxonomy + metadata).
# Reads metrics.json directly from disk, so it works even for encoders whose
# runs finished before the rest of the matrix does. Writes to a NEW directory
# per encoder (suffix "_phase2") — never touches lawgic_classifier_legal-bert_v3.
# Skips an encoder whose target directory already exists (idempotent / resumable),
# and skips an encoder with no completed dual-head run yet.

import shutil
from datetime import datetime, timezone

import torch
from safetensors.torch import load_file as load_safetensors
from transformers import AutoTokenizer

SAVE_TARGETS = {
    "nlpaueb/legal-bert-base-uncased": "legal-bert",
    "bert-base-uncased": "bert",
    "xlnet-base-cased": "xlnet",
    "roberta-base": "roberta",
}
SAVED_MODELS_DIR = PROJECT_ROOT / "saved_models"


def completed_dual_runs(encoder_name: str) -> list[dict]:
    records = []
    for metrics_path in sorted(tm.RUNS_DIR.glob("*/metrics.json")):
        record = json.loads(metrics_path.read_text())
        if record["encoder_name"] == encoder_name and record["heads"] == "dual":
            records.append(record)
    return records


def best_checkpoint_dir(run_id: str) -> Path:
    checkpoints = sorted(
        (tm.RUNS_DIR / run_id / "checkpoints").glob("checkpoint-*"),
        key=lambda p: int(p.name.split("-")[-1]),
    )
    if not checkpoints:
        raise FileNotFoundError(f"No checkpoint saved for {run_id}")
    # save_total_limit=1 + load_best_model_at_end=True: the one surviving
    # checkpoint is the best validation checkpoint, not just the last epoch.
    return checkpoints[-1]


def save_best_model(encoder_name: str, short_name: str) -> None:
    candidates = completed_dual_runs(encoder_name)
    if not candidates:
        print(f"skip {short_name}: no completed dual-head runs yet")
        return

    best = max(candidates, key=lambda r: r["best_val_metric"])
    run_id = best["run_id"]

    output_dir = SAVED_MODELS_DIR / f"lawgic_classifier_{short_name}_phase2"
    if output_dir.exists():
        print(f"skip {short_name}: {output_dir} already exists, not overwriting")
        return

    checkpoint_dir = best_checkpoint_dir(run_id)

    model = tm.LawgicDualHeadModel(encoder_name)
    weights_file = checkpoint_dir / "model.safetensors"
    state_dict = (
        load_safetensors(str(weights_file))
        if weights_file.exists()
        else torch.load(checkpoint_dir / "pytorch_model.bin", map_location="cpu", weights_only=True)
    )
    model.load_state_dict(state_dict)

    tokenizer = AutoTokenizer.from_pretrained(str(checkpoint_dir))

    output_dir.mkdir(parents=True)

    # Full state dict + encoder/tokenizer + heads separately, mirroring v3's layout.
    torch.save(model.state_dict(), output_dir / "model_state_dict.pt")
    model.encoder.save_pretrained(output_dir)
    tokenizer.save_pretrained(output_dir)
    torch.save(model.topic_head.state_dict(), output_dir / "topic_head_weights.pt")
    torch.save(model.harm_head.state_dict(), output_dir / "harm_head_weights.pt")

    topic_ids, name_by_topic, _ = core.load_taxonomy()
    compact_taxonomy = [
        {"classifier_id": i, "topic_id": tid, "name": name_by_topic[tid]}
        for i, tid in enumerate(topic_ids)
    ]
    (output_dir / "lawgic_topics_44.json").write_text(json.dumps(compact_taxonomy, indent=2))
    shutil.copy2(core.TAXONOMY_PATH, output_dir / "lawgic_topics_original_45.json")

    (output_dir / "test_metrics.json").write_text(json.dumps(best, indent=2, default=str))

    metadata = {
        "model_name": encoder_name,
        "architecture": "dual_head",
        "num_topics": core.NUM_LAWGIC_TOPICS,
        "num_harm_classes": core.NUM_HARM_CLASSES,
        "max_length": core.MAX_LENGTH,
        "decision_threshold": core.DECISION_THRESHOLD,
        "seed": best["seed"],
        "source_run_id": run_id,
        "best_val_metric": best["best_val_metric"],
        "seeds_considered": sorted(r["seed"] for r in candidates),
        "saved_at": datetime.now(timezone.utc).isoformat(),
        "note": (
            "Best-of-3-seeds model from the Phase 2 multi-encoder matrix "
            "(notebooks/evaluation/02_multiseed_encoder_runs.ipynb); does not "
            "replace lawgic_classifier_legal-bert_v3."
        ),
    }
    (output_dir / "training_metadata.json").write_text(json.dumps(metadata, indent=2))

    print(f"[{short_name}] saved best seed {best['seed']} (run {run_id}) -> {output_dir}")


for encoder_name, short_name in SAVE_TARGETS.items():
    save_best_model(encoder_name, short_name)


[legal-bert] saved best seed 2024 (run legal-bert-base-uncased__seed2024__dual) -> C:\Users\Enrique\Coding Projects\Thesis\lawgic\saved_models\lawgic_classifier_legal-bert_phase2
[bert] saved best seed 2024 (run bert-base-uncased__seed2024__dual) -> C:\Users\Enrique\Coding Projects\Thesis\lawgic\saved_models\lawgic_classifier_bert_phase2
[xlnet] saved best seed 2024 (run xlnet-base-cased__seed2024__dual) -> C:\Users\Enrique\Coding Projects\Thesis\lawgic\saved_models\lawgic_classifier_xlnet_phase2


Some weights of RobertaModel were not initialized from the model checkpoint at roberta-base and are newly initialized: ['pooler.dense.bias', 'pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


[roberta] saved best seed 2024 (run roberta-base__seed2024__dual) -> C:\Users\Enrique\Coding Projects\Thesis\lawgic\saved_models\lawgic_classifier_roberta_phase2


## Aggregation

Everything below reads the persisted run artifacts, so it can be re-run without a GPU.

In [13]:
run_files = sorted(tm.RUNS_DIR.glob("*/metrics.json"))
runs = pd.DataFrame([json.loads(p.read_text()) for p in run_files])
runs = runs[runs["holdout_source"].isna()] if "holdout_source" in runs else runs
print(f"Loaded {len(runs)} Phase 2 runs from {tm.RUNS_DIR}")
display(runs[["run_id", "encoder_name", "seed", "heads", "epochs_run", "wall_seconds", *core.HEADLINE_METRICS]])

runs.to_csv(core.EVAL_OUT_DIR / "phase2_runs.csv", index=False)
print(f"\nMeasured wall time: {runs['wall_seconds'].mean() / 60:.1f} min/run "
      f"(total {runs['wall_seconds'].sum() / 3600:.1f} h)")

Loaded 18 Phase 2 runs from C:\Users\Enrique\Coding Projects\Thesis\lawgic\generated_files\lawgic_taxonomy\runs


,run_id,encoder_name,seed,heads,epochs_run,wall_seconds,topic_macro_f1,topic_micro_f1,risk_accuracy,risk_macro_f1
0,bert-base-uncased__seed1337__dual,bert-base-uncased,1337,dual,15.0,2950.104609,0.768129,0.833610,0.826284,0.819565
1,bert-base-uncased__seed2024__dual,bert-base-uncased,2024,dual,20.0,3875.743355,0.780256,0.836836,0.836480,0.831212
2,bert-base-uncased__seed42__dual,bert-base-uncased,42,dual,20.0,3820.460032,0.775652,0.832867,0.826284,0.820721
3,legal-bert-base-uncased__seed1337__dual,nlpaueb/legal-bert-base-uncased,1337,dual,14.0,2716.261310,0.769291,0.832933,0.837991,0.832859
4,legal-bert-base-uncased__seed1337__risk,nlpaueb/legal-bert-base-uncased,1337,risk,6.0,1154.273496,NaN,NaN,0.823263,0.817329
5,legal-bert-base-uncased__seed1337__topic,nlpaueb/legal-bert-base-uncased,1337,topic,16.0,3073.803454,0.778805,0.837679,NaN,NaN
6,legal-bert-base-uncased__seed2024__dual,nlpaueb/legal-bert-base-uncased,2024,dual,20.0,8875.998135,0.770311,0.831424,0.836480,0.830932
7,legal-bert-base-uncased__seed2024__risk,nlpaueb/legal-bert-base-uncased,2024,risk,8.0,1534.044447,NaN,NaN,0.819109,0.815364
8,legal-bert-base-uncased__seed2024__topic,nlpaueb/legal-bert-base-uncased,2024,topic,13.0,2456.937220,0.767939,0.829530,NaN,NaN
9,legal-bert-base-uncased__seed42__dual,nlpaueb/legal-bert-base-uncased,42,dual,19.0,3825.444545,0.774095,0.836288,0.840634,0.834920



Measured wall time: 60.0 min/run (total 18.0 h)


In [14]:
grouped = runs.groupby(["encoder_name", "heads"])
aggregate = grouped[list(core.HEADLINE_METRICS)].agg(["mean", "std", "count"])
aggregate.columns = ["_".join(c) for c in aggregate.columns]
aggregate = aggregate.reset_index()
display(aggregate)
aggregate.to_csv(core.EVAL_OUT_DIR / "phase2_aggregate.csv", index=False)

,encoder_name,heads,topic_macro_f1_mean,topic_macro_f1_std,topic_macro_f1_count,topic_micro_f1_mean,topic_micro_f1_std,topic_micro_f1_count,risk_accuracy_mean,risk_accuracy_std,risk_accuracy_count,risk_macro_f1_mean,risk_macro_f1_std,risk_macro_f1_count
0,bert-base-uncased,dual,0.774679,0.006122,3,0.834438,0.002110,3,0.829683,0.005887,3,0.823833,0.006416,3
1,nlpaueb/legal-bert-base-uncased,dual,0.771232,0.002531,3,0.833548,0.002490,3,0.838369,0.002103,3,0.832904,0.001994,3
2,nlpaueb/legal-bert-base-uncased,risk,NaN,NaN,0,NaN,NaN,0,0.823389,0.004344,3,0.818690,0.004177,3
3,nlpaueb/legal-bert-base-uncased,topic,0.777473,0.008943,3,0.833381,0.004093,3,NaN,NaN,0,NaN,NaN,0
4,roberta-base,dual,0.775544,0.006427,3,0.837075,0.002948,3,0.841767,0.005192,3,0.836281,0.005201,3
5,xlnet-base-cased,dual,0.761874,0.030934,3,0.834009,0.006944,3,0.835599,0.006967,3,0.829607,0.007767,3


### Bootstrap CIs on the test metrics

Per run, 1,000 clause-level resamples of the test split, computed from the stored logits.
Reported alongside the across-seed sd: the bootstrap CI measures *test-set* sampling
noise, the sd measures *initialisation/ordering* noise. They are different quantities and
the manuscript should not conflate them.

In [15]:
N_RESAMPLES = 1000


def load_run_logits(run_id: str) -> dict:
    payload = np.load(tm.RUNS_DIR / run_id / "test_logits.npz")
    return {
        "topic_logits": payload["topic_logits"],
        "harm_logits": payload["harm_logits"],
        "arrays": {
            "labels": payload["labels"],
            "label_masks": payload["label_masks"],
            "harm_labels": payload["harm_labels"],
            "harm_masks": payload["harm_masks"],
        },
        "row_id": payload["row_id"],
    }


ci_rows = []
for run_id in runs["run_id"]:
    payload = load_run_logits(run_id)
    ci = core.bootstrap_ci(
        payload["topic_logits"], payload["harm_logits"], payload["arrays"], n_resamples=N_RESAMPLES
    )
    ci.insert(0, "run_id", run_id)
    ci_rows.append(ci)

bootstrap_table = pd.concat(ci_rows, ignore_index=True)
bootstrap_table.to_csv(core.EVAL_OUT_DIR / "phase2_bootstrap_ci.csv", index=False)
display(bootstrap_table.head(12))

,run_id,metric,point,mean,ci_low,ci_high,n_rows
0,bert-base-uncased__seed1337__dual,topic_macro_f1,0.768129,0.757783,0.729956,0.782843,2648
1,bert-base-uncased__seed1337__dual,topic_micro_f1,0.833610,0.833595,0.819431,0.846433,2648
2,bert-base-uncased__seed1337__dual,risk_accuracy,0.826284,0.826367,0.811169,0.841399,2648
3,bert-base-uncased__seed1337__dual,risk_macro_f1,0.819565,0.819525,0.803590,0.835370,2648
4,bert-base-uncased__seed2024__dual,topic_macro_f1,0.780256,0.769983,0.739307,0.797077,2648
5,bert-base-uncased__seed2024__dual,topic_micro_f1,0.836836,0.836801,0.822386,0.850479,2648
6,bert-base-uncased__seed2024__dual,risk_accuracy,0.836480,0.836613,0.822508,0.850831,2648
7,bert-base-uncased__seed2024__dual,risk_macro_f1,0.831212,0.831249,0.816272,0.846257,2648
8,bert-base-uncased__seed42__dual,topic_macro_f1,0.775652,0.764637,0.734771,0.790399,2648
9,bert-base-uncased__seed42__dual,topic_micro_f1,0.832867,0.832821,0.818347,0.846218,2648


### Paired significance tests

Both tests are **paired on the clause**: every run scored the identical test rows in the
identical order, so a difference is attributable to the varied component and nothing else.

- **Risk head — McNemar.** Item-level correctness per clause (over `harm_mask=1` rows),
  exact binomial on the discordant pairs. This is the right test for two classifiers on
  one sample; an unpaired accuracy comparison would throw away the pairing and lose power.
- **Topic head — paired bootstrap.** Macro-F1 is not decomposable into per-item
  correctness, so McNemar does not apply. Instead each resample draws one set of clause
  indices and scores *both* models on it; the reported interval is over the difference.

Seeds are averaged out by comparing the **best seed** of each arm; change `pick` below to
compare a fixed seed if you would rather not condition on validation performance.

In [ ]:
def best_run(encoder: str, heads: str = "dual") -> str:
    subset = runs[(runs["encoder_name"] == encoder) & (runs["heads"] == heads)]
    if subset.empty:
        raise KeyError(f"no runs for {encoder}/{heads}")
    return subset.sort_values("best_val_metric", ascending=False).iloc[0]["run_id"]


def compare(run_a: str, run_b: str) -> dict:
    a, b = load_run_logits(run_a), load_run_logits(run_b)
    assert np.array_equal(a["row_id"], b["row_id"]), "runs were scored on different rows"
    arrays = a["arrays"]

    valid = arrays["harm_masks"].astype(bool)
    correct_a = a["harm_logits"][valid].argmax(1) == arrays["harm_labels"][valid]
    correct_b = b["harm_logits"][valid].argmax(1) == arrays["harm_labels"][valid]
    mcnemar = core.mcnemar(correct_a, correct_b)

    def delta(indices):
        ma = core.topic_metrics(a["topic_logits"][indices], arrays["labels"][indices], arrays["label_masks"][indices])
        mb = core.topic_metrics(b["topic_logits"][indices], arrays["labels"][indices], arrays["label_masks"][indices])
        return ma["topic_macro_f1"] - mb["topic_macro_f1"]

    paired = core.paired_bootstrap_delta(delta, np.arange(len(arrays["labels"])), n_resamples=N_RESAMPLES)
    return {
        "run_a": run_a,
        "run_b": run_b,
        "risk_mcnemar_b": mcnemar["b"],
        "risk_mcnemar_c": mcnemar["c"],
        "risk_mcnemar_p": mcnemar["p_value"],
        "topic_macro_f1_delta": paired["delta"],
        "topic_delta_ci_low": paired["ci_low"],
        "topic_delta_ci_high": paired["ci_high"],
        "topic_delta_p": paired["p_value"],
    }


LEGAL_BERT = tm.ENCODERS[0]
comparisons = []
for other in tm.ENCODERS[1:]:
    try:
        comparisons.append(compare(best_run(LEGAL_BERT), best_run(other)))
    except KeyError as exc:
        print(f"skipped: {exc}")

# Head ablation, original scope: dual vs each single-head variant, legal-bert only.
# Kept verbatim so the Section 4.4.3 numbers stay traceable to the cell that produced them.
for heads in ("topic", "risk"):
    try:
        comparisons.append(compare(best_run(LEGAL_BERT), best_run(LEGAL_BERT, heads)))
    except KeyError as exc:
        print(f"skipped: {exc}")

# Head ablation, extended scope: the same two comparisons for the other three encoders,
# same bootstrap / McNemar apparatus, no new test. Legal-BERT is skipped here because the
# loop above already appended it.
ablation_rows = []
for encoder in tm.ENCODERS:
    for heads in ("topic", "risk"):
        try:
            row = compare(best_run(encoder, "dual"), best_run(encoder, heads))
        except KeyError as exc:
            print(f"skipped: {exc}")
            continue
        ablation_rows.append({"encoder_name": encoder, "heads": heads, **row})
        if encoder != LEGAL_BERT:
            comparisons.append(row)

ablation = pd.DataFrame(ablation_rows)
ablation.to_csv(core.EVAL_OUT_DIR / "phase2_head_ablation.csv", index=False)
display(ablation)

significance = pd.DataFrame(comparisons)
significance.to_csv(core.EVAL_OUT_DIR / "phase2_significance.csv", index=False)
display(significance)

## Output table (a) — headline, rows = encoder/config

Cells are `mean +/- sd` over seeds. `n/a` marks a metric a configuration cannot produce:
topic-only leaves the risk head untrained, risk-only leaves the topic head untrained, so
reporting those cells would be reporting random weights.

In [ ]:
LABELS = {
    "topic_macro_f1": "Topic macro-F1",
    "topic_micro_f1": "Topic micro-F1",
    "risk_accuracy": "Risk accuracy",
    "risk_macro_f1": "Risk macro-F1",
}
SHORT_NAMES = {
    "nlpaueb/legal-bert-base-uncased": "Legal-BERT",
    "bert-base-uncased": "BERT",
    "xlnet-base-cased": "XLNet",
    "roberta-base": "RoBERTa",
}
HEAD_NAMES = {"dual": "dual", "topic": "topic-only", "risk": "risk-only"}
# 4 encoders x 3 head configs, encoder-major so each encoder's three rows sit together.
CONFIG_NAMES = {
    (encoder, heads): f"{SHORT_NAMES[encoder]} ({HEAD_NAMES[heads]})"
    for encoder in tm.ENCODERS
    for heads in ("dual", "topic", "risk")
}


def mean_sd(values: pd.Series) -> str:
    if values.isna().all():
        return "n/a"
    return f"{values.mean():.3f} ± {values.std(ddof=1):.3f}" if len(values) > 1 else f"{values.mean():.3f}"


headline = pd.DataFrame(
    [
        {
            "Configuration": CONFIG_NAMES.get((encoder, heads), f"{encoder} ({heads})"),
            "Seeds": int(group["seed"].nunique()),
            **{LABELS[m]: mean_sd(group[m]) for m in core.HEADLINE_METRICS},
        }
        for (encoder, heads), group in runs.groupby(["encoder_name", "heads"])
    ]
)
order = [CONFIG_NAMES[k] for k in CONFIG_NAMES if CONFIG_NAMES[k] in set(headline["Configuration"])]
headline = headline.set_index("Configuration").loc[order].reset_index()
display(headline)

core.write_outputs(
    headline,
    "phase2_headline",
    caption=(
        "Test performance by encoder and head configuration, mean $\\pm$ standard deviation "
        "over three seeds (42, 1337, 2024), for the full 4 encoder x 3 head-config design. "
        "All runs use the identical persisted seed-42 "
        "clause split and identical hyperparameters; only the encoder, the seed and the "
        "active heads vary."
    ),
    label="tab:encoder-matrix",
)

## Cross-encoder consistency of the dual-head benefit

The point of running the ablation on all four encoders is not four more rows of numbers —
it is whether the four deltas *agree*. Two views of the same quantity are reported side by
side, and neither is collapsed into a grand mean:

- **Seed-level delta.** Per encoder, dual-head minus single-head on the metric that
  single-head arm still produces (risk macro-F1 for dual-vs-risk-only, topic macro-F1 for
  dual-vs-topic-only), paired by seed, reported as mean +/- sd over the three seed pairs.
  With n = 3 the interval shown is mean +/- t(0.975, 2) * sd / sqrt(3) — wide by
  construction, and that width is the honest statement of what three seeds buy.
- **Best-seed delta.** The McNemar b/c/p (risk) and paired-bootstrap CI (topic) already
  computed in the ablation cell above, on the best-validation seed of each arm. Same
  machinery as every other comparison in this notebook and in the chapter.

The check reported at the end: do all four seed-level intervals overlap a common value,
and does any single encoder's interval sit clear of the other three's range. Overlap
supports "the dual-head benefit is a property of the training objective"; a non-overlapping
encoder rejects it and the claim must be scoped to the backbones where it holds.

In [ ]:
# Metric each single-head arm can still be compared on. topic-only leaves the risk head
# untrained and risk-only leaves the topic head untrained, so each contrast has exactly one
# valid metric — the same "n/a" logic as the headline table.
ABLATION_METRIC = {"risk": "risk_macro_f1", "topic": "topic_macro_f1"}
T_CRIT = 4.303  # t(0.975, df=2): three seeds, two-sided 95%


def seed_level_delta(encoder: str, heads: str, metric: str) -> dict:
    """dual minus single-head on `metric`, paired seed by seed."""
    subset = runs[runs["encoder_name"] == encoder]
    dual = subset[subset["heads"] == "dual"].set_index("seed")[metric]
    single = subset[subset["heads"] == heads].set_index("seed")[metric]
    seeds = sorted(set(dual.index) & set(single.index))
    if not seeds:
        raise KeyError(f"no paired seeds for {encoder} dual vs {heads}")
    deltas = np.array([dual[s] - single[s] for s in seeds], dtype=float)
    half_width = T_CRIT * deltas.std(ddof=1) / np.sqrt(len(deltas)) if len(deltas) > 1 else np.nan
    return {
        "encoder": SHORT_NAMES[encoder],
        "contrast": f"dual - {HEAD_NAMES[heads]}",
        "metric": metric,
        "n_seeds": len(deltas),
        "delta_mean": deltas.mean(),
        "delta_sd": deltas.std(ddof=1) if len(deltas) > 1 else np.nan,
        "ci_low": deltas.mean() - half_width,
        "ci_high": deltas.mean() + half_width,
        "all_seeds_positive": bool((deltas > 0).all()),
        "per_seed": np.round(deltas, 4).tolist(),
    }


def overlap_report(table: pd.DataFrame, title: str) -> None:
    """Flag whether the per-encoder intervals share a common value, and name any outlier."""
    print(f"\n{title}")
    for _, row in table.iterrows():
        print(f"  {row['encoder']:<11} {row['delta_mean']:+.4f} +/- {row['delta_sd']:.4f} "
              f"[{row['ci_low']:+.4f}, {row['ci_high']:+.4f}]  seeds={row['per_seed']}")

    lower, upper = table["ci_low"].max(), table["ci_high"].min()
    if lower <= upper:
        print(f"  -> all {len(table)} intervals overlap on [{lower:+.4f}, {upper:+.4f}]: "
              "consistent with one common effect.")
    else:
        # No common value. Name the encoders whose interval is disjoint from every other's.
        outliers = [
            row["encoder"] for _, row in table.iterrows()
            if all(row["ci_high"] < o["ci_low"] or row["ci_low"] > o["ci_high"]
                   for _, o in table.iterrows() if o["encoder"] != row["encoder"])
        ]
        print("  -> NO common value: the four intervals do not share a point.")
        print(f"     disjoint from all others: {outliers or 'none individually — pairwise only'}")

    signs = set(np.sign(table["delta_mean"]))
    print(f"     sign agreement: {'all same sign' if len(signs) == 1 else 'SIGNS DISAGREE'}"
          f" ({', '.join(f'{r.encoder} {r.delta_mean:+.4f}' for r in table.itertuples())})")


consistency_rows = []
for heads, metric in ABLATION_METRIC.items():
    for encoder in tm.ENCODERS:
        try:
            consistency_rows.append(seed_level_delta(encoder, heads, metric))
        except KeyError as exc:
            print(f"skipped: {exc}")

consistency = pd.DataFrame(consistency_rows)

# Attach the best-seed significance already computed above, so the seed-level spread and
# the paired test sit in one table instead of two.
if not ablation.empty:
    keyed = ablation.set_index([ablation["encoder_name"].map(SHORT_NAMES),
                                ablation["heads"].map(lambda h: f"dual - {HEAD_NAMES[h]}")])
    for column in ("risk_mcnemar_b", "risk_mcnemar_c", "risk_mcnemar_p",
                   "topic_macro_f1_delta", "topic_delta_ci_low", "topic_delta_ci_high",
                   "topic_delta_p"):
        consistency[column] = [
            keyed[column].get((row.encoder, row.contrast), np.nan) for row in consistency.itertuples()
        ]
    # Blank the columns that do not apply to a contrast: McNemar is a risk-head test, the
    # paired bootstrap a topic-head one.
    risk_rows = consistency["metric"] == "risk_macro_f1"
    consistency.loc[risk_rows, ["topic_macro_f1_delta", "topic_delta_ci_low",
                                "topic_delta_ci_high", "topic_delta_p"]] = np.nan
    consistency.loc[~risk_rows, ["risk_mcnemar_b", "risk_mcnemar_c", "risk_mcnemar_p"]] = np.nan

consistency.to_csv(core.EVAL_OUT_DIR / "phase2_head_ablation_consistency.csv", index=False)
display(consistency)

risk_side = consistency[consistency["metric"] == "risk_macro_f1"].reset_index(drop=True)
topic_side = consistency[consistency["metric"] == "topic_macro_f1"].reset_index(drop=True)
if len(risk_side) > 1:
    overlap_report(risk_side, "Risk macro-F1: dual-head minus risk-only, per encoder")
if len(topic_side) > 1:
    overlap_report(topic_side, "Topic macro-F1: dual-head minus topic-only, per encoder")


## Output table (b) — per-topic breakdown for the final model

Rows = 44 topics + macro avg + weighted avg, columns = precision / recall / F1 / support.
Reported twice: for the **best legal-bert seed** (what a deployed single model achieves)
and as the **mean across the three seeds** (what the architecture achieves). Support is
identical across seeds because the split is.

In [18]:
topic_ids, name_by_topic, _ = core.load_taxonomy()

best_legal_bert = best_run(LEGAL_BERT)
best_table = pd.read_csv(tm.RUNS_DIR / best_legal_bert / "per_topic.csv")

seed_tables = [
    pd.read_csv(tm.RUNS_DIR / run_id / "per_topic.csv").set_index("topic_id")
    for run_id in runs[(runs["encoder_name"] == LEGAL_BERT) & (runs["heads"] == "dual")]["run_id"]
]
mean_table = sum(t[["precision", "recall", "f1"]] for t in seed_tables) / len(seed_tables)
mean_table = mean_table.join(seed_tables[0][["support", "observed"]]).reset_index()

per_topic = best_table.merge(mean_table, on="topic_id", suffixes=("_best", "_mean"))
per_topic.insert(1, "topic_name", per_topic["topic_id"].map(lambda t: name_by_topic.get(t, t)))
display(per_topic)

core.write_outputs(
    per_topic[["topic_id", "topic_name", "precision_best", "recall_best", "f1_best",
               "f1_mean", "support_best"]].rename(columns={
        "topic_id": "Topic", "topic_name": "Name", "precision_best": "P", "recall_best": "R",
        "f1_best": "F1", "f1_mean": "F1 (seed mean)", "support_best": "Support"}),
    "phase2_per_topic",
    caption=(
        f"Per-topic test performance of the best Legal-BERT dual-head seed ({best_legal_bert}), "
        "with the mean F1 across the three seeds for comparison. Support counts supervised "
        "positive cells in the test split; topics with no observed test cells are omitted."
    ),
    label="tab:per-topic",
)

,topic_id,topic_name,precision_best,recall_best,f1_best,support_best,observed_best,precision_mean,recall_mean,f1_mean,support_mean,observed_mean
0,choice_of_law,Choice of Law,0.925926,0.840336,0.881057,119,2358,0.912351,0.871148,0.890967,119,2358
1,choice_of_forum,Choice of Forum,0.908257,0.825000,0.864629,120,2360,0.905022,0.872222,0.887903,120,2360
2,mandatory_arbitration,Mandatory Arbitration,0.886792,0.839286,0.862385,56,2357,0.901638,0.875000,0.888003,56,2357
3,class_action_waiver,Class Action Waiver,0.892857,0.877193,0.884956,57,2357,0.914402,0.871345,0.892135,57,2357
4,limitation_of_liability,Limitation of Liability,0.882883,0.837607,0.859649,117,2438,0.885705,0.794872,0.837446,117,2438
5,liability_cap,Liability Cap,0.444444,0.571429,0.500000,7,2336,0.448148,0.428571,0.431624,7,2336
6,warranty_disclaimer,Warranty Disclaimer,0.896970,0.902439,0.899696,164,2336,0.912626,0.902439,0.907183,164,2336
7,indemnification,Indemnification,0.000000,0.000000,0.000000,0,2336,0.000000,0.000000,0.000000,0,2336
8,limitation_period,Limitation Period,1.000000,1.000000,1.000000,1,141,1.000000,1.000000,1.000000,1,141
9,contract_changes,Contract Changes,0.843750,0.900000,0.870968,150,2377,0.842885,0.882222,0.862064,150,2377


(WindowsPath('C:/Users/Enrique/Coding Projects/Thesis/lawgic/generated_files/lawgic_taxonomy/evaluation/phase2_per_topic.csv'),
 WindowsPath('C:/Users/Enrique/Coding Projects/Thesis/lawgic/generated_files/lawgic_taxonomy/evaluation/phase2_per_topic.tex'))